src/components/fields/file-upload/file-upload.tsx

```diff
@@ -0,0 +1,241 @@
+import { FileUpload as DSFileUpload, FileItemProps } from "@lifesg/react-design-system/file-upload";
+import { useContext, useEffect, useRef } from "react";
+import { useFormContext } from "react-hook-form";
+import * as Yup from "yup";
+import { IGenericFieldProps } from "..";
+import { FileHelper } from "../../../utils";
+import { useFieldEvent, useValidationConfig } from "../../../utils/hooks";
+import { IYupValidationRule } from "../../frontend-engine";
+import { ERROR_MESSAGES, Sanitize } from "../../shared";
+import { FileUploadContext, FileUploadProvider } from "./file-upload-context";
+import { FileUploadHelper } from "./file-upload-helper";
+import { FileUploadManager } from "./file-upload-manager";
+import { EFileStatus, IFile, IFileUploadSchema, IFileUploadValidationRule, IFileUploadValue } from "./types";
+
+export const FileUploadInner = (props: IGenericFieldProps<IFileUploadSchema>) => {
+	// =============================================================================
+	// CONST, STATE, REF
+	// =============================================================================
+	const {
+		id,
+		error,
+		isDirty,
+		value,
+		schema: { description, label, uploadOnAddingFile, validation, warning, ...otherSchema },
+	} = props;
+	const { files, setFiles } = useContext(FileUploadContext);
+	const fileTypeRuleRef = useRef<IFileUploadValidationRule>({});
+	const maxFilesRuleRef = useRef<IYupValidationRule>({});
+	const maxFileSizeRuleRef = useRef<IFileUploadValidationRule>({});
```

Comment:
just curious, what's the use of putting these rules in a ref? setting ref doesn't trigger rerenders, although for this case it works because it's caused from a change in `validation` prop, which sets state and thus guarantees to trigger a rerender anyway

Reply:
I had to change to refs because at the time of initialisation, right after the rules are defined in the `useEffect`, they are needed as part of handling default values. because the states are not getting updated fast enough, I switched to refs to get around this issue.